plugins/tpm/fu-tpm-device.c

```diff
@@ -0,0 +1,291 @@
+				     "failed to initialize TPM");
+		return FALSE;
+	}
+
+	/* lookup guaranteed details from TPM */
+	family = fu_plugin_tpm_get_string (ctx, TPM2_PT_FAMILY_INDICATOR, error);
+	if (family == NULL) {
+		g_prefix_error (error, "failed to read TPM family");
+		return FALSE;
+	}
+	manufacturer = fu_plugin_tpm_get_string (ctx, TPM2_PT_MANUFACTURER, error);
+	if (manufacturer == NULL) {
+		g_prefix_error (error, "failed to read TPM manufacturer");
+		return FALSE;
+	}
+	model1 = fu_plugin_tpm_get_string (ctx, TPM2_PT_VENDOR_STRING_1, error);
+	if (model1 == NULL) {
+		g_prefix_error (error, "failed to read TPM vendor string");
+		return FALSE;
+	}
+	if (!fu_plugin_tpm_get_uint32 (ctx, TPM2_PT_VENDOR_TPM_TYPE, &tpm_type, error)) {
+		g_prefix_error (error, "failed to read TPM type");
+		return FALSE;
+	}
+
+	/* these are not guaranteed by spec and may be NULL */
+	model2 = fu_plugin_tpm_get_string (ctx, TPM2_PT_VENDOR_STRING_2, error);
+	model3 = fu_plugin_tpm_get_string (ctx, TPM2_PT_VENDOR_STRING_3, error);
+	model4 = fu_plugin_tpm_get_string (ctx, TPM2_PT_VENDOR_STRING_4, error);
+	model = g_strjoin ("", model1, model2, model3, model4, NULL);
+
+	/* add GUIDs to daemon */
+	id1 = g_strdup_printf ("TPM\\VEN_%s&DEV_%04X", manufacturer, tpm_type);
+	fu_device_add_instance_id (device, id1);
+	id2 = g_strdup_printf ("TPM\\VEN_%s&MOD_%s", manufacturer, model);
+	fu_device_add_instance_id (device, id2);
+	id3 = g_strdup_printf ("TPM\\VEN_%s&DEV_%04X&VER_%s", manufacturer, tpm_type, family);
+	fu_device_add_instance_id (device, id3);
+	id4 = g_strdup_printf ("TPM\\VEN_%s&MOD_%s&VER_%s", manufacturer, model, family);
+	fu_device_add_instance_id (device, id4);
+
+	/* enforce vendors can only ship updates for thier own hardware */
+	vendor_id = g_strdup_printf ("TPM:%s", manufacturer);
+	fu_device_set_vendor_id (device, vendor_id);
+	tmp = fu_tpm_device_convert_manufacturer (manufacturer);
+	fu_device_set_vendor (device, tmp != NULL ? tmp : manufacturer);
+
+	/* get version */
+	if (!fu_plugin_tpm_get_uint32 (ctx, TPM2_PT_FIRMWARE_VERSION_1, &version1, error))
+		return FALSE;
+	if (!fu_plugin_tpm_get_uint32 (ctx, TPM2_PT_FIRMWARE_VERSION_2, &version2, error))
+		return FALSE;
+	version_raw = ((guint64) version1) << 32 | ((guint64) version2);
+	fu_device_set_version_raw (device, version_raw);
+
+	/* this has to be done after _add_instance_id() adds the quirks */
+	verfmt = fu_device_get_version_format (device);
+	version = fu_common_version_from_uint64 (version_raw, verfmt);
+	fu_device_set_version (device, version, verfmt);
+
+	/* try to add the event log as a child */
+	if (!fu_tpm_device_setup_event_log (self, &error_local))
+		g_warning ("no event log: %s", error_local->message);
+
+	/* success */
+	return TRUE;
+}
+
+static void
+fu_tpm_device_init (FuTpmDevice *self)
+{
+	fu_device_set_name (FU_DEVICE (self), "TPM");
+	fu_device_set_version_format (FU_DEVICE (self), FWUPD_VERSION_FORMAT_QUAD);
```

Comment:
I think you probably also want to add some flags for internal too right?

Reply:
Agree, thanks.